### PR TITLE
Don't autoremember flags

### PR DESCRIPTION
**Pull request by smlance**
Based on https://github.com/bundler/bundler/pull/3744.

### CODE REVIEW

**Comment by smlance** [comment]
(Waiting for https://github.com/bundler/bundler/pull/3939 to be merged.)

**Comment by smlance** [comment]
It looks like certain matchers in `matchers.rb`, e.g., `should_be_installed`, rely on settings configured in previous invocations of `bundle`. This makes certain specs that use `bundle` commands with flags fail, because these settings are not persisted to the matcher commands. Thoughts about how to deal with this? @indirect, @segiddins?

**Comment by segiddins** [comment]
@smlance allow passing flags to `should_be_installed`?

**Comment by indirect** [comment]
I think it makes sense to allow `path` as an argument to `should_be_installed`... is there anything else we need to support? As an alternative to an argument, we could also potentially create a more generic block form for config settings, something like `config(:path => "/wherever") { should_be_installed("rack 1.0") }`

File: lib/bundler/settings.rb [hunk 1]
```diff
@@ -176,11 +185,14 @@ def groups_array(group_type, scope)
     # Always returns an absolute path to the bundle directory
     # TODO: Refactor this method
     def path
+      # TODO: Shoudl we consider the @current_config here?
```
Comment (segiddins) [inline lib/bundler/settings.rb]: why should / shouldn't we?

File: lib/bundler/settings.rb [hunk 2]
```diff
@@ -260,8 +274,22 @@ def get_array(key)
...
-      self[key] = (array.empty? ? nil : array.join(":")) if array
+    def set_array(key, array, scope)
+      if scope != :current and scope != :local and scope != :global
+        Bundler.ui.error "Invalid scope #{scope} given. Please use :local or :global."
+        exit 1
+      end
+
+      if array
+        value = (array.empty? ? nil : array.join(":"))
+        if scope == :current
```
Comment (segiddins) [inline lib/bundler/settings.rb]: please use a `case` statement for this, and do the scope checking in there

File: spec/commands/clean_spec.rb [hunk 3]
```diff
@@ -470,12 +472,19 @@ def should_not_have_gems(*gems)
...
     G
     bundle "install --system"
-    bundle "clean --force"
 
-    expect(out).to include("Removing foo (1.0)")
-    sys_exec "gem list"
-    expect(out).not_to include("foo (1.0)")
-    expect(out).to include("rack (1.0.0)")
+    # See cli/install.rb:L76.
+    # FIXME: [user-unfriendly] `clean` only happens if path is set to
```
Comment (segiddins) [inline spec/commands/clean_spec.rb]: this looks like a bug that needs to get fixed

File: spec/commands/config_spec.rb [hunk 4]
```diff
@@ -14,9 +14,12 @@
       bundle "config path vendor/bundle"
       bundle "install"
 
-      expect(bundled_app(".bundle")).not_to exist
-      expect(tmp("foo/bar/config")).to exist
-      should_be_installed "rack 1.0.0"
+      # See cli/install.rb:L86
```
Comment (segiddins) [inline spec/commands/config_spec.rb]: please don't lint to lines -- use methods, which are less likely to go out of date

File: lib/bundler/settings.rb (hunk 1 above)
Comment (smlance) [inline lib/bundler/settings.rb]: Never mind -- I don't think we need to. The comment just below it explains why: `--path` is no longer used, so `bundle install --path foo` or any command that previously accepted `--path` would now be set with `bundle config`, which only set options in the `local` and `global` scopes.

File: lib/bundler/settings.rb [hunk 5]
```diff
@@ -9,15 +9,15 @@ class Settings
     attr_reader :root
 
     def initialize(root = nil)
-      @root          = root
-      @local_config  = load_config(local_config_file)
-      @global_config = load_config(global_config_file)
+      @root           = root
+      @current_config ||= {}
```
Comment (segiddins) [inline lib/bundler/settings.rb]: no need to `||=` in `initialize`

File: lib/bundler/settings.rb [hunk 6]
```diff
@@ -260,8 +289,20 @@ def get_array(key)
       self[key] ? self[key].split(":").map(&:to_sym) : []
     end
 
-    def set_array(key, array)
-      self[key] = (array.empty? ? nil : array.join(":")) if array
+    def set_array(key, array, scope)
+      unless [:current, :local, :global].include? scope
+        Bundler.ui.error "Invalid scope #{scope} given. Please use :local or :global."
```
Comment (segiddins) [inline lib/bundler/settings.rb]: raise an exception?

File: lib/bundler/settings.rb [hunk 7]
```diff
@@ -160,11 +185,13 @@ def with(scope = nil)
...
+      when nil
         get_array(group_type)
-      elsif scope == :global
+      when :global
         @global_config[key] ? @global_config[key].split(" ").map(&:to_sym) : []
-      elsif scope == :local
+      when :local
         @local_config[key] ? @local_config[key].split(" ").map(&:to_sym) : []
       else
         Bundler.ui.error "Invalid scope #{scope} given. Please use :local or :global."
```
Comment (segiddins) [inline lib/bundler/settings.rb]: raise an exception here?

File: spec/support/builders.rb [hunk 8]
```diff
@@ -365,7 +365,6 @@ def build_with(builder, name, args, &blk)
       versions = args.last || "1.0"
       spec     = nil
 
-      # TODO: [2.0] Change this, since we won't use options[:path]
       options[:path] ||= @_build_path
```
Comment (segiddins) [inline spec/support/builders.rb]: shouldn't we change this?

File: spec/support/helpers.rb [hunk 9]
```diff
@@ -1,5 +1,13 @@
 module Spec
   module Helpers
+    def set_temp_config(config)
+      if block_given?
+        config.each {|k, v| Bundler.settings.set_local(k, v) }
+        yield
+        config.each {|k, _| Bundler.settings.set_local(k, nil) }
```
Comment (segiddins) [inline spec/support/helpers.rb]: shouldn't this restore the settings?

Comment (smlance) [inline spec/support/helpers.rb]: Oh -- good point.

File: spec/support/builders.rb (hunk 8 above)
Comment (smlance) [inline spec/support/builders.rb]: Hm. I'll look into this later.

Comment (smlance) [inline spec/support/builders.rb]: No. It looks like this options hash is specific to this file and has nothing to do with the options hash passed to `bundle` commands.

Comment (segiddins) [inline spec/support/builders.rb]: ok, so the `TODO` was incorrect?

Comment (smlance) [inline spec/support/builders.rb]: Yes. I added it incorrectly.

File: lib/bundler/cli/config.rb [hunk 10]
```diff
@@ -109,7 +112,17 @@ def run
...
+        Bundler.ui.warn "`path` is already configured, so it will be unset."
+        delete_config("path")
+        # We still want `system` to be set.
+        :no_conflict
+      elsif name == "path" and Bundler.settings[:system]
+        Bundler.ui.warn "`system` is already configured, so it will be unset."
+        delete_config("system")
+        # We still want `path` to be set.
+        :no_conflict
+      elsif name == "path.system" and Bundler.settings[:path] and new_value == "true"
```
Comment (indirect) [inline lib/bundler/cli/config.rb]: Oops. We wound up with `path`, `system`, and `path.system`. I intended us to only have two. :) I think `path` and `path.system` is the clearest. Let's get rid of just plain `system`.

File: spec/support/helpers.rb [hunk 11]
```diff
@@ -1,5 +1,18 @@
 module Spec
   module Helpers
+    def set_temp_config(config)
```
Comment (indirect) [inline spec/support/helpers.rb]: I think renaming this to `with_config` would match with Ruby conventions for methods that take a block and then unset after.

File: spec/realworld/parallel_spec.rb [hunk 12]
```diff
@@ -9,7 +9,9 @@
       gem 'i18n', '~> 0.6.0' # Because 0.7+ requires Ruby 1.9.3+
     G
 
-    bundle :install, :jobs => 4, :env => { "DEBUG" => "1" }
+    bundle "config jobs 4"
+    # NOTE: The env option below cannot be configured with `bundle "config env debug"`
+    bundle :install, :env => { "DEBUG" => "1" }
```
Comment (indirect) [inline spec/realworld/parallel_spec.rb]: it's actually possible to just use `bundle install --verbose` here (verbose output contains the thread numbers, which is why we use verbose output in this test)

File: lib/bundler/settings.rb [hunk 13]
```diff
@@ -30,13 +31,17 @@ def [](name)
...
 
-    def []=(key, value)
+    def set_current(key, value)
+      @current_config[key_for(key)] = value
+    end
+
+    alias_method :[]=, :set_current
+
+    def set_local(key, value)
       local_config_file or raise GemfileNotFound, "Could not locate #{SharedHelpers.gemfile_name}"
```
Comment (segiddins) [inline lib/bundler/settings.rb]: please don't use `or`

File: spec/support/helpers.rb [hunk 14]
```diff
@@ -1,5 +1,18 @@
...
+    def with_config(config)
+      if block_given?
+        old_config = {}
+
+        config.each do |k, v|
+          old_config[k] = Bundler.settings[k]
+          Bundler.settings.set_local(k, v)
+        end
+        yield
+        config.each {|k, _| Bundler.settings.set_local(k, old_config[k]) }
```
Comment (segiddins) [inline spec/support/helpers.rb]: this needs to be in an `ensure`

**Comment by segiddins** [comment]
This seems good?

File: spec/support/helpers.rb (hunk 14 above)
Comment (smlance) [inline spec/support/helpers.rb]: Which part?

File: lib/bundler/settings.rb (hunk 13 above)
Comment (smlance) [inline lib/bundler/settings.rb]: Would `unless` be preferable?

**Comment by segiddins** [comment]
@smlance you'll need to squash the `wip` commits out, they don't provide any information

**Comment by smlance** [comment]
Now that I removed the flags from the list of `method_option`s, `thor` detects that the flags are invalid before `bundler` does, so the error messages I added won't ever be shown.

Should each command have its own `path` setting? They all shared the same `path` configuration before, and this seemed to only be a problem for `binstubs` (see the fixed commits in https://github.com/smlance/bundler/commit/fe9e4246804ecab4b6a5b98f563371b8358625c8). Based on the specs, this isn't a problem for the other commands, so I'm okay with allowing them to share it.

File: lib/bundler/settings.rb (hunk 13 above)
Comment (segiddins) [inline lib/bundler/settings.rb]: Yes

File: spec/support/helpers.rb (hunk 14 above)
Comment (segiddins) [inline spec/support/helpers.rb]: Restoring the old value

**Comment by indirect** [comment]
This looks good!

**Comment by indirect** [comment]
@homu r+

**Pull request merged**